triggers/inotify/inotify.go

```diff
@@ -0,0 +1,308 @@
+	for _, p := range d.paths {
+		if strings.HasPrefix(path, p.Path) {
+			return &p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("path object not found: %v", path)
+}
+
+func (d *daemon) worker() {
+	// close watcher
+	defer d.watcher.Close()
+
+	// process events
+	for {
+		select {
+		case event := <-d.watcher.Events:
+			// new filesystem event
+			d.log.Trace().
+				Interface("event", event).
+				Msg("Filesystem event")
+
+			switch {
+			case event.Op&fsnotify.Create == fsnotify.Create:
+				// create
+				fi, err := os.Stat(event.Name)
+				if err != nil {
+					d.log.Error().
+						Err(err).
+						Str("path", event.Name).
+						Msg("Failed retrieving filesystem info")
+					continue
+				}
+
+				// watch new directories
+				if fi.IsDir() {
+					if err := filepath.Walk(event.Name, d.walkFunc); err != nil {
+						d.log.Error().
+							Err(err).
+							Str("path", event.Name).
+							Msg("Failed watching new directory")
+					}
+
+					continue
+				}
+
+			case event.Op&fsnotify.Rename == fsnotify.Rename, event.Op&fsnotify.Remove == fsnotify.Remove:
+				// renamed / removed
+			default:
+				// ignore this event
+				continue
+			}
+
+			// get path object
+			p, err := d.getPathObject(event.Name)
+			if err != nil {
+				d.log.Error().
+					Err(err).
+					Str("path", event.Name).
+					Msg("Failed determining path object")
+				continue
+			}
+
+			// rewrite
+			rewritten := p.Rewriter(event.Name)
+
+			// filter
+			if !p.Allowed(rewritten) {
+				continue
+			}
+
+			// get directory where path has an extension
+			if filepath.Ext(rewritten) != "" {
+				// there was most likely a file extension, use the directory
+				rewritten = filepath.Dir(rewritten)
+			}
+
+			// move to queue
+			d.queue.inputs <- rewritten
+
+		case err := <-d.watcher.Errors:
+			d.log.Error().
+				Err(err).
+				Msg("Failed receiving filesystem events")
+		}
+	}
+}
+
+type queue struct {
+	callback autoscan.ProcessorFunc
+	log      zerolog.Logger
+	priority int
+	inputs   chan string
+	scans    map[string]time.Time
+	lock     *sync.Mutex
+}
+
+func newQueue(cb autoscan.ProcessorFunc, log zerolog.Logger, priority int) *queue {
+	q := &queue{
+		callback: cb,
+		log:      log,
+		priority: priority,
+		inputs:   make(chan string),
+		scans:    make(map[string]time.Time),
+		lock:     &sync.Mutex{},
+	}
+
+	go q.worker()
+
+	return q
+}
+
+func (q *queue) add(path string) {
+	// acquire lock
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	// queue scan task
+	q.scans[path] = time.Now().Add(10 * time.Second)
+}
+
+func (q *queue) worker() {
+	for {
+		select {
+		case path, ok := <-q.inputs:
+			if !ok {
+				// channel closed
+				return
+			}
+
+			// add path to queue
+			q.add(path)
+
+		default:
+			// process queue
+			q.process()
+		}
+	}
+}
+
+func (q *queue) process() {
+	// acquire lock
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	// sleep if no scans queued
+	if len(q.scans) == 0 {
+		time.Sleep(100 * time.Millisecond)
+		return
+	}
```

Comment:
sleep locks the mutex

Reply:
thats fine though, as the channel can continue to receive queued "paths", which will be processed on unlock